p2p/net/swarm/swarm.go

```diff
@@ -310,6 +321,13 @@ func (s *Swarm) addConn(tc transport.CapableConn, dir network.Direction) (*Conn,
 	})
 	c.notifyLk.Unlock()
 
+	if isFirstConn {
```

Comment:
Are we sure we want this outside of the `notifyLk`? That changes the behavior from what it used to be (inside the notifyAll).

Reply:
Yeah, I think this needs to be inside. We need to make sure that the disconnect comes after the connect event.

Reply:
Hm. Actually, I'm not sure if this lock helps us here. We need to make this guarantee across all connections.